### PR TITLE
PR: Don't wait for disconnected comm (IPython console)

### LINKED ISSUE #13288: TimeoutError: Timeout while waiting for comm port.
### Issue Report Checklist

* [X] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ ] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [X] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description

this is a duplication of [#11889](https://github.com/spyder-ide/spyder/issues/11889) but as that one is closed, and the problem still persists, @goanpeca asked me to report this one next time it pops up.

The issue reporter however didn't have the `Submit to Github` button enabled (not sure if that is a bug in itself) anyway, I'll do it on Github itself :wink: 

### What steps reproduce the problem?
On a Linux (Ubuntu 20.04) box, installed [miniforge](https://github.com/conda-forge/miniforge) (so everything comes from conda-forge) and then I created a fresh environment like this:
```shell
(base) me@mybox:~$ conda create -n s4 python=3.7
(base) me@mybox:~$ conda activate s4
(s4) me@mybox:~$ conda install spyder
...
(s4) me@mybox:~$ conda list spyder
# packages in environment at /home/me/my-envs/s4:
#
# Name                    Version                   Build  Channel
spyder                    4.1.3                    py37_0  
spyder-kernels            1.9.1                    py37_0  
(s4) me@mybox:~$ spyder
libpng warning: bKGD: invalid
libpng warning: bKGD: invalid
libpng warning: bKGD: invalid
libpng warning: bKGD: invalid
```
the `libpng` warnings are probably not harmfull, but good to know I would say :wink: 
anyway, after a while the `issue reporter` pops up (best case), sometimes it doesn't pop up, but one can see the exception on the internal console! (not sure what that is all about)

### What is the expected output? What do you see instead?
N/A

### Paste Traceback/Error Below (if applicable)

```python-traceback
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/qtconsole/base_frontend_mixin.py", line 138, in _dispatch
    handler(msg)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/widgets/namespacebrowser.py", line 172, in _handle_execute_reply
    self.refresh_namespacebrowser()
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/widgets/namespacebrowser.py", line 63, in refresh_namespacebrowser
    callback=self.set_namespace_view
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder_kernels/comms/commbase.py", line 538, in __call__
    call_dict, call_data, self._comm_id)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 130, in _get_call_return_value
    with self.comm_channel_manager(comm_id, queue_message=not interrupt):
  File "/home/nerohmot/my-envs/s4/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 87, in comm_channel_manager
    timeout)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 168, in _wait
    raise TimeoutError(timeout_msg)
TimeoutError: Timeout while waiting for comm port.
Traceback (most recent call last):
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/qtconsole/base_frontend_mixin.py", line 138, in _dispatch
    handler(msg)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/widgets/namespacebrowser.py", line 172, in _handle_execute_reply
    self.refresh_namespacebrowser()
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/widgets/namespacebrowser.py", line 63, in refresh_namespacebrowser
    callback=self.set_namespace_view
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder_kernels/comms/commbase.py", line 538, in __call__
    call_dict, call_data, self._comm_id)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 130, in _get_call_return_value
    with self.comm_channel_manager(comm_id, queue_message=not interrupt):
  File "/home/nerohmot/my-envs/s4/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 87, in comm_channel_manager
    timeout)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 168, in _wait
    raise TimeoutError(timeout_msg)
TimeoutError: Timeout while waiting for comm port.
Traceback (most recent call last):
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/plugin.py", line 554, in set_current_client_working_directory
    shellwidget.set_cwd(directory)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/widgets/shell.py", line 178, in set_cwd
    self.call_kernel(interrupt=True).set_cwd(dirname)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder_kernels/comms/commbase.py", line 538, in __call__
    call_dict, call_data, self._comm_id)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 130, in _get_call_return_value
    with self.comm_channel_manager(comm_id, queue_message=not interrupt):
  File "/home/nerohmot/my-envs/s4/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 87, in comm_channel_manager
    timeout)
  File "/home/nerohmot/my-envs/s4/lib/python3.7/site-packages/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 168, in _wait
    raise TimeoutError(timeout_msg)
TimeoutError: Timeout while waiting for comm port.
```

## Versions

* Spyder version: 4.1.3
* Python version: 3.7.6
* Qt version: 5.9.7
* PyQt version: 5.9.2
* Operating System name/version: Linux Ubuntu 20.04 (x86_64)

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0           :  1.4.0 (OK)
chardet >=2.0.0                :  3.0.4 (OK)
cloudpickle >=0.5.0            :  1.5.0 (OK)
diff_match_patch >=20181111    :  20181111 (OK)
intervaltree                   :  None (OK)
IPython >=4.0                  :  7.16.1 (OK)
jedi =0.15.2                   :  0.15.2 (OK)
keyring                        :  None (OK)
nbconvert >=4.0                :  5.6.1 (OK)
numpydoc >=0.6.0               :  1.1.0 (OK)
parso =0.5.2                   :  0.5.2 (OK)
pexpect >=4.4.0                :  4.8.0 (OK)
pickleshare >=0.4              :  0.7.5 (OK)
psutil >=5.3                   :  5.7.0 (OK)
pygments >=2.0                 :  2.6.1 (OK)
pylint >=0.25                  :  2.5.3 (OK)
pyls >=0.31.9;<0.32.0          :  0.31.10 (OK)
qdarkstyle >=2.8               :  2.8.1 (OK)
qtawesome >=0.5.7              :  0.7.2 (OK)
qtconsole >=4.6.0              :  4.7.5 (OK)
qtpy >=1.5.0                   :  1.9.0 (OK)
rtree >=0.8.3                  :  0.9.4 (OK)
sphinx >=0.6.6                 :  3.1.2 (OK)
spyder_kernels >=1.9.1;<1.10.0 :  1.9.1 (OK)
watchdog                       :  None (OK)
xdg >=0.26                     :  0.26 (OK)
zmq >=17                       :  19.0.1 (OK)

# Optional:
cython >=0.21                  :  None (OK)
matplotlib >=2.0.0             :  None (OK)
numpy >=1.7                    :  None (OK)
pandas >=0.13.1                :  None (OK)
scipy >=0.17.0                 :  None (OK)
sympy >=0.7.3                  :  None (OK)
```

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes
Stop waiting on disconnected comms. Instead:
If interrupting: don't interrupt and send via regular channel.
If blocking: raise an error and let the calling function handle the error

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #13288

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

File: spyder/plugins/ipythonconsole/comms/kernelcomm.py [hunk 1]
```diff
@@ -144,10 +140,19 @@ def _get_call_return_value(self, call_dict, call_data, comm_id):
...
         interrupt = 'interrupt' in settings and settings['interrupt']
-
+        # Need to make sure any blocking call is replied rapidely.
+        if not self.comm_channel_connected():
+            # Ask again for comm config
+            self.remote_call()._send_comm_config()
+            # Can not interrupt if comm not connected
+            interrupt = False
+            logger.info("Dropping interrupt because comm is disconnected: ",
+                            str(call_dict))
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/comms/kernelcomm.py:150]: Please fix the indentation for this line.

File: spyder/plugins/ipythonconsole/widgets/help.py [hunk 2]
```diff
@@ -131,25 +132,25 @@ def is_defined(self, objtxt, force_import=False):
         """Return True if object is defined"""
         try:
             return self.call_kernel(
-                interrupt=True, blocking=True
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/help.py:134]: Will these methods work while on debugging? I mean, I thought it was necessary to interrupt the kernel to make these calls while debugging.

**Review by ccordoba12 — COMMENTED** [review]
This looks pretty good, thanks @impact27!

My only concern is that you're removing `interrupt=True` from a lot of calls, which I thought was necessary while debugging.

File: spyder/plugins/ipythonconsole/widgets/help.py (hunk 2 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/help.py:134]: I changed it so that blocking implies interrupt

**Comment by impact27** [comment]
> This looks pretty good, thanks @impact27!
> 
> My only concern is that you're removing `interrupt=True` from a lot of calls, which I thought was necessary while debugging.

All the calls that use `blocking=True` need `interrupt=True`, so I just changed it so `blocking` implies `interrupt`.

File: spyder/plugins/ipythonconsole/widgets/help.py [hunk 3]
```diff
@@ -131,25 +132,25 @@ def is_defined(self, objtxt, force_import=False):
...
-        except (TimeoutError, UnpicklingError, RuntimeError):
+        except (TimeoutError, UnpicklingError, RuntimeError, CommError):
             return None
 
     def get_doc(self, objtxt):
         """Get object documentation dictionary"""
         try:
-            return self.call_kernel(interrupt=True, blocking=True
+            return self.call_kernel(blocking=True
                                     ).get_doc(objtxt)
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/help.py:144]:
```suggestion
            return self.call_kernel(blocking=True).get_doc(objtxt)
```

File: spyder/plugins/ipythonconsole/widgets/help.py [hunk 4]
```diff
@@ -131,25 +132,25 @@ def is_defined(self, objtxt, force_import=False):
...
-        except (TimeoutError, UnpicklingError, RuntimeError):
+        except (TimeoutError, UnpicklingError, RuntimeError, CommError):
             return None
 
     def get_source(self, objtxt):
         """Get object source"""
         try:
-            return self.call_kernel(interrupt=True, blocking=True
+            return self.call_kernel(blocking=True
                                     ).get_source(objtxt)
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/help.py:152]:
```suggestion
            return self.call_kernel(blocking=True).get_source(objtxt)
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks a lot @impact27 for your help with this! I left some simple suggestions for you and clarification question for me.

File: spyder/plugins/ipythonconsole/comms/kernelcomm.py [hunk 5]
```diff
@@ -140,11 +136,27 @@ def _get_call_return_value(self, call_dict, call_data, comm_id):
...
                 return
 
         settings = call_dict['settings']
         interrupt = 'interrupt' in settings and settings['interrupt']
-
+        interrupt = interrupt or blocking
+        # Need to make sure any blocking call is replied rapidely.
+        if interrupt and not self.comm_channel_connected():
+            # Ask again for comm config
+            self.remote_call()._send_comm_config()
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/ipythonconsole/comms/kernelcomm.py:151]: Maybe the config was not received for some reason, so I ask again so the next call might not fail. It won't help for this call though.

**Comment by ccordoba12** [comment]
@impact27, is this one ready?

**Comment by impact27** [comment]
@ccordoba12 yes

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27! Let's hope this problem is finally fixed by your work in this PR.

**Pull request merged**